File: backend/MarketWorkplace.Domain/Entities/Subscription.cs
using System.ComponentModel.DataAnnotations;

namespace MarketWorkplace.Domain.Entities;

/// <summary>A plan a user is subscribed to — powers the dashboard Subscriptions page
/// and is shown on the linked mobile user's profile.</summary>
public class Subscription
{
    /// <summary>Billing cadences supported by the platform.</summary>
    public const string Monthly = "Monthly";
    public const string Yearly = "Yearly";

    /// <summary>Known plans.</summary>
    public const string BasicPlan = "Basic";
    public const string PremiumPlan = "Premium";
    public const string EnterprisePlan = "Enterprise";

    /// <summary>Active subscription states.</summary>
    public const string Active = "Active";
    public const string Inactive = "Inactive";
    public const string Expired = "Expired";

    public int Id { get; set; }

    /// <summary>Foreign key of the subscribed <see cref="User"/>.</summary>
    public int UserId { get; set; }

    /// <summary>The subscribed account (mobile users primarily, dashboard admins too).</summary>
    public User User { get; set; } = null!;

    /// <summary>Plan name, e.g. <c>Basic</c>, <c>Premium</c> or <c>Enterprise</c>.</summary>
    [Required, StringLength(30)]
    public string Plan { get; set; } = BasicPlan;

    /// <summary>Monthly price in USD for this plan.</summary>
    [Range(0, 1_000_000)]
    public decimal Price { get; set; }

    /// <summary>How often the plan bills.</summary>
    [Required, StringLength(10)]
    public string BillingCycle { get; set; } = Monthly;

    /// <summary>Whether the subscription is currently usable.</summary>
    [Required, StringLength(10)]
    public string Status { get; set; } = Active;

    /// <summary>When the subscription started.</summary>
    public DateTime StartsAt { get; set; } = DateTime.UtcNow;

    /// <summary>When the subscription ends, or <c>null</c> while it is open-ended.</summary>
    public DateTime? EndsAt { get; set; }

    /// <summary>Whether it renews automatically at the end of the cycle.</summary>
    public bool AutoRenew { get; set; }

    /// <summary>When the subscription was created.</summary>
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
}

File: mobile/lib/features/profile/subscription.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../core/api/api_client.dart';
import '../auth/auth_controller.dart';

/// A plan bound to an account, as returned by `GET /api/subscriptions/user/{id}`.
///
/// The mobile app only ever reads its own rows: the dashboard's
/// `/subscriptions` page is where plans are created, edited and deleted.
class UserSubscription {
  const UserSubscription({
    required this.id,
    required this.userId,
    required this.plan,
    required this.price,
    required this.billingCycle,
    required this.status,
    required this.startsAt,
    required this.endsAt,
    required this.autoRenew,
  });

  final int id;
  final int userId;

  /// `Basic`, `Premium` or `Enterprise`.
  final String plan;

  /// Price in USD for one billing cycle.
  final double price;

  /// `Monthly` or `Yearly`.
  final String billingCycle;

  /// `Active`, `Inactive` or `Expired`.
  final String status;

  final DateTime startsAt;

  /// `null` for an open-ended plan.
  final DateTime? endsAt;

  final bool autoRenew;

  factory UserSubscription.fromJson(Map<String, dynamic> json) =>
      UserSubscription(
        id: (json['id'] as num?)?.toInt() ?? 0,
        userId: (json['userId'] as num?)?.toInt() ?? 0,
        plan: json['plan'] as String? ?? '',
        price: (json['price'] as num?)?.toDouble() ?? 0,
        billingCycle: json['billingCycle'] as String? ?? '',
        status: json['status'] as String? ?? '',
        startsAt: DateTime.tryParse(json['startsAt'] as String? ?? '') ??
            DateTime.now(),
        endsAt: DateTime.tryParse(json['endsAt'] as String? ?? ''),
        autoRenew: json['autoRenew'] as bool? ?? false,
      );
}

/// The signed-in account's plans — an empty list when it has none.
///
/// Re-fetches whenever the session changes so a different user never sees the
/// previous account's plan.
final mySubscriptionsProvider =
    FutureProvider.autoDispose<List<UserSubscription>>((ref) async {
  final session = ref.watch(authControllerProvider).value;
  if (session == null) return const <UserSubscription>[];
  final dio = ref.watch(apiClientProvider);
  final response = await guardApi(
    () => dio.get<List<dynamic>>('/api/subscriptions/user/${session.user.id}'),
  );
  return [
    for (final row in response.data ?? const <dynamic>[])
      UserSubscription.fromJson(row as Map<String, dynamic>),
  ];
});
